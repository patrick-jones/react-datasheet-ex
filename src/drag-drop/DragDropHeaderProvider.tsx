import {Component, ReactNode} from 'react';

import {InjectedDragDropProps, DragDropHeaderProps} from './interfaces';
import {HeaderDragSource, HeaderDropTarget} from './header-dnd';


export interface DragDropHeaderProviderProps
  extends DragDropHeaderProps,
    InjectedDragDropProps {
  children: (props: InjectedDragDropProps) => ReactNode;
}


class DragDropHeader extends Component<DragDropHeaderProviderProps> {

  render(): ReactNode {
    const {children, ...rest} = this.props;
    return children(rest);
  }
}

/**
 * Provides all drag-and-drop properties to a render function assigned to the
 * `children` property.
 */
const DragDropHeaderProvider = HeaderDropTarget(HeaderDragSource(DragDropHeader));

export default DragDropHeaderProvider;
