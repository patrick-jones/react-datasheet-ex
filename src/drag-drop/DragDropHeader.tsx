import React, {Component, ReactNode} from 'react';

import {ColumnHeader} from '../interfaces';
import HeaderTitle from '../HeaderTitle';
import {
  HEADER_DRAG_SOURCE,
  HEADER_DROP_TARGET,
  HEADER_CELL_CONTAINER,
} from '../ClassNames';

import {InjectedDragDropProps, DragDropHeaderRendererProps} from './interfaces';
import DragDropHeaderProvider from './DragDropHeaderProvider';


export default class DragDropHeader<T extends ColumnHeader = ColumnHeader>
  extends Component<DragDropHeaderRendererProps<T>> {

  renderHeader = (dndProps: InjectedDragDropProps) => {
    const {connectDragSource, connectDropTarget, isDropOver} = dndProps;

    const classNames = [
      HEADER_DRAG_SOURCE,
      isDropOver ? HEADER_DROP_TARGET : '',
    ].filter(s => !!s).join(' ');

    return connectDropTarget(connectDragSource(
      <span className={classNames}>
        <HeaderTitle {...this.props} />
      </span>
    ));
  };

  render(): ReactNode {
    const {col, onHeaderDropped} = this.props;

    return (
      <DragDropHeaderProvider col={col} onHeaderDropped={onHeaderDropped}>
        {this.renderHeader}
      </DragDropHeaderProvider>
    );
  }
}
