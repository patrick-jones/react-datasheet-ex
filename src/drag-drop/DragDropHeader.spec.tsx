import React from 'react';
import {mount} from 'enzyme';

import {DragDropContext} from 'react-dnd';
import Backend, {TestBackend} from 'react-dnd-test-backend';
import toJson from 'enzyme-to-json';

import DragDropHeader from './DragDropHeader';
import DragDropHeaderProvider from './DragDropHeaderProvider';


function testableComponent(spyA: jest.Mock, spyB: jest.Mock) {
  return DragDropContext(Backend)(

    class TestContainer extends React.Component {
      render() {
        return (
          <table>
            <thead>
              <tr>
                <td>
                  <DragDropHeader
                    col={0}
                    header={{id: 'A', title: 'A'}}
                    onHeaderDropped={spyA}
                  >
                    A
                  </DragDropHeader>
                </td>
                <td>
                  <DragDropHeader
                    col={1}
                    header={{id: 'B', title: 'B'}}
                    onHeaderDropped={spyB}
                  >
                    B
                  </DragDropHeader>
                </td>
              </tr>
            </thead>
          </table>
        );
      }
    }
  );
}


describe('DragDropHeader', () => {
  it('matches snapshot', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Calls onHeaderDropped', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    // console.log(wrapper.debug());
    // @ts-ignore
    const backend: TestBackend = wrapper.instance().getManager().getBackend();

    const hp = wrapper.find(DragDropHeaderProvider);

    // @ts-ignore
    const dragSource = hp.at(1).childAt(0).instance().getHandlerId();
    // @ts-ignore
    const dropTarget = hp.at(0).instance().getHandlerId();

    backend.simulateBeginDrag([dragSource]);
    backend.simulateHover([dropTarget]);
    backend.simulateDrop();
    backend.simulateEndDrag();

    expect(spyA.mock.calls.length).toBe(1);
    expect(spyB.mock.calls.length).toBe(0);

    expect(spyA.mock.calls[0][0]).toBe(1);
    expect(spyA.mock.calls[0][1]).toBe(0);
  });
});
