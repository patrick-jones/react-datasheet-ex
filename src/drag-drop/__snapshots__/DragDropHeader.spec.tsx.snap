// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDropHeader matches snapshot 1`] = `
<DragDropContext(TestContainer)>
  <TestContainer>
    <table>
      <thead>
        <tr>
          <td>
            <DragDropHeader
              col={0}
              header={
                Object {
                  "id": "A",
                  "title": "A",
                }
              }
              onHeaderDropped={[MockFunction]}
            >
              <DropTarget(DragSource(DragDropHeader))
                col={0}
                onHeaderDropped={[MockFunction]}
              >
                <DragSource(DragDropHeader)
                  col={0}
                  connectDropTarget={[Function]}
                  isDropOver={false}
                  onHeaderDropped={[MockFunction]}
                >
                  <DragDropHeader
                    col={0}
                    connectDragPreview={[Function]}
                    connectDragSource={[Function]}
                    connectDropTarget={[Function]}
                    isDropOver={false}
                    onHeaderDropped={[MockFunction]}
                  >
                    <span
                      className="rdx-header-drag-source"
                    >
                      <HeaderTitle
                        className=""
                        col={0}
                        header={
                          Object {
                            "id": "A",
                            "title": "A",
                          }
                        }
                        onHeaderDropped={[MockFunction]}
                      >
                        <span
                          className="rdx-header-cell-title rdx-header-cell-title--clip"
                        >
                          A
                        </span>
                      </HeaderTitle>
                    </span>
                  </DragDropHeader>
                </DragSource(DragDropHeader)>
              </DropTarget(DragSource(DragDropHeader))>
            </DragDropHeader>
          </td>
          <td>
            <DragDropHeader
              col={1}
              header={
                Object {
                  "id": "B",
                  "title": "B",
                }
              }
              onHeaderDropped={[MockFunction]}
            >
              <DropTarget(DragSource(DragDropHeader))
                col={1}
                onHeaderDropped={[MockFunction]}
              >
                <DragSource(DragDropHeader)
                  col={1}
                  connectDropTarget={[Function]}
                  isDropOver={false}
                  onHeaderDropped={[MockFunction]}
                >
                  <DragDropHeader
                    col={1}
                    connectDragPreview={[Function]}
                    connectDragSource={[Function]}
                    connectDropTarget={[Function]}
                    isDropOver={false}
                    onHeaderDropped={[MockFunction]}
                  >
                    <span
                      className="rdx-header-drag-source"
                    >
                      <HeaderTitle
                        className=""
                        col={1}
                        header={
                          Object {
                            "id": "B",
                            "title": "B",
                          }
                        }
                        onHeaderDropped={[MockFunction]}
                      >
                        <span
                          className="rdx-header-cell-title rdx-header-cell-title--clip"
                        >
                          B
                        </span>
                      </HeaderTitle>
                    </span>
                  </DragDropHeader>
                </DragSource(DragDropHeader)>
              </DropTarget(DragSource(DragDropHeader))>
            </DragDropHeader>
          </td>
        </tr>
      </thead>
    </table>
  </TestContainer>
</DragDropContext(TestContainer)>
`;
